server/src/main/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardSnapshots.java

```diff
@@ -208,6 +208,7 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         builder.endArray();
         // Then we list all snapshots with list of all blobs that are used by the snapshot
         builder.startObject(Fields.SNAPSHOTS);
+        assert shardSnapshots.stream().map(SnapshotFiles::snapshot).distinct().count() == shardSnapshots.size();
```

Comment:
I think this assertion would fail in cases where the repository has already ended up with two snapshots of the same name somehow. We can't have read that from an `index-N` file, but we fall back to listing the individual snapshots so I think it could still happen. In that case I think we shouldn't be writing this file (so I agree with this assertion) but we should also be checking our behaviour with a duplicated snapshot name, and logging warnings instead of trying to write this file at all.

Reply:
True, in theory, I agree. But in practice, I'm not so sure I'd want to backport that kind of change to `6.7` (we'd have to somehow do the logging and checking outside of `BlobStoreIndexShardSnapshots` and that's not going to be a clean change that will look the same in `6.7` and `7.x`).

I was gonna look into that part of the problem next, we shouldn't be indexing this stuff by snapshot name in the first place, because we're precisely using UUIDs to get around the possible race conditions that can lead to duplicate entries here. I think it makes more sense fixing the format here and adding the mechanics of handling existing duplicates into that effort?

Reply:
Right, I'm also ok with removing this assertion if we can't easily put those checks in place. I'm just against using an assertion to catch that we're operating on a broken repo (in a way in which we know it can be).

Throwing a proper exception here would also be better than writing an unreadable metadata file - at least this way the user would know to take some action.

Also +1 to improving this format in the near future.

Reply:
> Throwing a proper exception here would also be better than writing an unreadable metadata file - at least this way the user would know to take some action.

On reflection, I'm not sure this is true. That might leave the repo with a readable-but-stale metadata file; at least if the file is unreadable then we fall back to something sensible.

Reply:
> Right, I'm also ok with removing this assertion if we can't easily put those checks in place. I'm just against using an assertion to catch that we're operating on a broken repo (in a way in which we know it can be).

I'd like tests to fail here if that's ok? I can see a few sequences of events that lead us here but are pretty tricky to reproduce/fix with the way snapshotting works currently, but I think it would be good to at least have a little coverage on this situation in tests?

Reply:
I would rather not use an assertion here to say that we haven't ended up in this state in a test. This assertion effectively says that it is not possible to be in this state, but we know that this isn't the case. It's an _unintended_ state, certainly, but we know it can happen.

Reply:
Fair point, killed the assertion in 443a5849988 :)